packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3070,12 +3070,33 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
       return null;
     }
     if (child != null) {
-      if (child.widget == newWidget) {
+      bool canUpdate = true;
+      // When the type of a widget is changed between Stateful and Stateless via
+      // hot reload, the element tree will end up in a partially invalid state.
+      // That is, if the widget was a StatefulWidget and is now a StatelessWidget,
+      // then the element tree currently contains a StatefulElement that is incorrectly
+      // referencing a StatelessWidget (and likewise with StatelessElement).
+      //
+      // To avoid crashing due to type errors, we need to gently guide the invalid
+      // element out of the tree. To do so, we ensure that the `canUpdate` condition
+      // returns false which prevents us from trying to update the existing element
+      // incorrectly.
+      //
+      // For the case where the widget becomes Stateful, we also need to avoid
+      // accessing `StatelessElement.widget` as the cast on the getter will
+      // cause a type error to be thrown. Here we avoid that by short-circuiting
+      // the `Widget.canUpdate` check once `canUpdate` is false.
+      assert(() {
+        canUpdate = !((child is StatelessElement && newWidget is StatefulWidget) ||
+                      (child is StatefulElement && newWidget is StatelessWidget));
+        return true;
+      }());
+      if (canUpdate && child.widget == newWidget) {
```

Comment:
maybe wrap both if statements in a single if statement that checks this? rather than having the condition in both?

Reply:
Done